Tools/scripts/run_coverage.py

```diff
@@ -0,0 +1,275 @@
+        self.INFO_FILE = os.path.join(root_dir, self.REPORT_DIR, "lcov.info")
+        self.INFO_FILE_BASE = os.path.join(root_dir, self.REPORT_DIR, "lcov_base.info")
+        self.LCOV_LOG = os.path.join(root_dir, "GCOV_lcov.log")
+        self.GENHTML_LOG = os.path.join(root_dir, "GCOV_genhtml.log")
+        self.CI = os.environ.get("CI", False)
+
+    def progress(self, string):
+        """Pretty printer."""
+        print("****** %s" % (string,))
+
+    def init_coverage(self):
+        """Initialize ArduPilot for coverage.
+
+        This need to be run with the binaries builded.
+        """
+        self.progress("Initilizing Coverage...")
+        self.progress("Removing previous reports")
+        try:
+            shutil.rmtree(self.REPORT_DIR)
+        except OSError as e:
+            print("Error: %s : %s" % (self.REPORT_DIR, e.strerror))
+
+        try:
+            os.makedirs(self.REPORT_DIR)
+        except OSError as error:
+            print(error)
+
+        self.progress("Checking that vehicles binaries are set up and built")
+        examples_dir = os.path.join(root_dir, 'build/linux/examples')
+        binaries_dir = os.path.join(root_dir, 'build/sitl/bin')
+        is_build_correct = self.check_build("example", examples_dir) and self.check_build("binaries", binaries_dir)
+        if not is_build_correct:
+            self.run_build()
+
+        self.progress("Zeroing previous build")
+        retcode = subprocess.call(["lcov", "--zerocounters", "--directory", root_dir])
+        if retcode != 0:
+            self.progress("Failed with retcode (%s)" % retcode)
+
+        self.progress("Initializing Coverage with current build")
+        try:
+            result = subprocess.run(["lcov", "--no-external", "--initial", "--capture",
+                                     "--exclude", "\"" + root_dir + "/modules/uavcan/*\"",
+                                     "--exclude", "\"" + root_dir + "/build/sitl/modules/*\"",
+                                     "--directory", root_dir,
+                                     "-o", self.INFO_FILE_BASE,
+                                     ], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, check=True)
+            if not self.CI:
+                print(result.stdout)
+            with open(self.LCOV_LOG, 'w') as log_file:
+                log_file.write(result.stdout)
+        except subprocess.CalledProcessError as err:
+            print("ERROR :")
+            print(err.cmd)
+            print(err.output)
+            exit(1)
+        self.progress("Initialization done !")
+
+    def check_build(self, name, path):
+        """Check that build directory is not empty and that binaries are built with the coverage flags."""
+        self.progress("Checking that %s are set up and built" % name)
+        if os.path.exists(path):
+            if not os.listdir(path):
+                return False
+        else:
+            return False
```

Comment:
```suggestion
        if not os.listdir(path):
            return False
```

You're not going to be able to list it if it doesn't exist.

Reply:
yep but I don't want listdir to raise FileNotFoundError exception !

Reply:
Then 
```suggestion
        try:
            os.listdir(path)
        except os.FileNotFoundError:
            pass
```

Your code can actually already raise that exception if there's a race condition.